src/node_http2.cc

```diff
@@ -199,27 +210,166 @@ Http2Priority::Http2Priority(Environment* env,
+  nghttp2_session_callbacks_set_error_callback(
+    callbacks, OnNghttpError);
+
+  if (kHasGetPaddingCallback) {
+    nghttp2_session_callbacks_set_select_padding_callback(
+      callbacks, OnSelectPadding);
+  }
+}
+
+
+Http2Session::Callbacks::~Callbacks() {
+  nghttp2_session_callbacks_del(callbacks);
+}
+
+
 Http2Session::Http2Session(Environment* env,
                            Local<Object> wrap,
                            nghttp2_session_type type)
     : AsyncWrap(env, wrap, AsyncWrap::PROVIDER_HTTP2SESSION),
-      StreamBase(env) {
+      session_type_(type) {
   MakeWeak<Http2Session>(this);
 
   Http2Options opts(env);
 
+  int32_t maxHeaderPairs = opts.GetMaxHeaderPairs();
+  max_header_pairs_ =
+      type == NGHTTP2_SESSION_SERVER
+          ? std::max(maxHeaderPairs, 4)     // minimum # of request headers
+          : std::max(maxHeaderPairs, 1);    // minumum # of response headers
+
+  max_outstanding_pings_ = opts.GetMaxOutstandingPings();
+
   padding_strategy_ = opts.GetPaddingStrategy();
 
-  int32_t maxHeaderPairs = opts.GetMaxHeaderPairs();
-  maxHeaderPairs = type == NGHTTP2_SESSION_SERVER ?
-      std::max(maxHeaderPairs, 4) : std::max(maxHeaderPairs, 1);
-  Init(type, *opts, nullptr, maxHeaderPairs);
+  bool hasGetPaddingCallback =
+      padding_strategy_ == PADDING_STRATEGY_MAX ||
+      padding_strategy_ == PADDING_STRATEGY_CALLBACK;
+
+  nghttp2_session_callbacks* callbacks
+      = callback_struct_saved[hasGetPaddingCallback ? 1 : 0].callbacks;
+
+  typedef int (*init_fn)(nghttp2_session** session,
+                         const nghttp2_session_callbacks* callbacks,
+                         void* user_data,
+                         const nghttp2_option* options);
+  init_fn fn = type == NGHTTP2_SESSION_SERVER ?
+      nghttp2_session_server_new2 :
+      nghttp2_session_client_new2;
+
+  // This should fail only if the system is out of memory, which
+  // is going to cause lots of other problems anyway, or if any
+  // of the options are out of acceptable range, which we should
+  // be catching before it gets this far. Either way, crash if this
+  // fails.
+  CHECK_EQ(fn(&session_, callbacks, this, *opts), 0);
+
+  Start();
+}
+
+
+Http2Session::~Http2Session() {
+  CHECK(persistent().IsEmpty());
+  Close();
+}
 
-  // For every node::Http2Session instance, there is a uv_prepare_t handle
-  // whose callback is triggered on every tick of the event loop. When
-  // run, nghttp2 is prompted to send any queued data it may have stored.
+// For every node::Http2Session instance, there is a uv_prepare_t handle
+// whose callback is triggered on every tick of the event loop. When
+// run, nghttp2 is prompted to send any queued data it may have stored.
+// TODO(jasnell): Currently, this creates one uv_prepare_t per Http2Session,
+//                we should investigate to see if it's faster to create a
+//                single uv_prepare_t for all Http2Sessions, then iterate
+//                over each.
```

Comment:
Just so you know what I’m having in mind for this: I’d keep a boolean flag on a each session that keeps track of whether a send callback is scheduled to be executed, and use the `SetImmediate()` from #17117 to schedule it once it becomes necessary (if that flag is not already set).

In particular I’d really like to avoid iterating over *all* sessions.

Reply:
whether or not there is data to send is not always easily visible. nghttp has a `want_write` function that can be checked to see if a particular session has data to send.

Reply:
nghttp2 doesn't do it's own I/O so it's always something that *we* do that makes it want to write something, right? We should use `want_write` after such an action to check.

(Fwiw, I would be up for trying to do this myself once this PR is merged, and if I prove myself wrong, so be it. ;))

Reply:
I'm not saying it's a bad idea :-) ... just saying that it's not always immediately visible.

Btw... Currently, we are already calling `SendPendingData()` immediately after receiving and processing data on a socket. It's entirely possible that we could eliminate the prep_t call entirely, I just haven't worked on that yet.

Reply:
To expand a bit... the underlying nghttp2_session will have data to write to the output stream in one of two cases:

1. We receive data from the socket and pass it in to nghttp2_session. This will trigger a series of synchronous callbacks to user code. In some cases, nghttp2 will push it's own data onto the send queue during this time, in other cases user code will push it's own data onto the send queue. Either way, when we're doing receiving the chunk of data, we call `SendPendingData()` to push out whatever was sent.

2. User-code takes other actions independent of receiving data from the socket. The `uv_prep_t` is really in place to handle this second set of actions. Rather than calling `SendPendingData()` multiple times, once after each action, the data is allowed to batch up in nghttp2 internal queue and is sent on each tick of the event loop.